components/script_bindings/str.rs

```diff
@@ -293,6 +298,62 @@ impl DOMString {
     }
 }
 
+/// Because this converts to a DOMString it become UTF-8 encoded which is closer to
+/// the spec definition of <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-json-bytes>
+/// but we generally do not operate on anything that is truly a WTF-16 string.
+///
+/// <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-a-json-string>
+pub fn serialize_jsval_to_json_utf8(
+    cx: SafeJSContext,
+    jsval: HandleValue,
+) -> Result<DOMString, Error> {
+    #[repr(C)]
+    struct VoidableString {
+        is_void: bool,
+        data: String,
+    }
+
+    let mut out_str = VoidableString {
+        is_void: true,
+        data: String::new(),
+    };
+
+    #[allow(unsafe_code)]
+    unsafe extern "C" fn stringified(
+        string: *const u16,
+        len: u32,
+        data: *mut std::ffi::c_void,
+    ) -> bool {
+        let s = data as *mut VoidableString;
+        let string_chars = slice::from_raw_parts(string, len as usize);
+        (*s).is_void = false;
+        (*s).data.push_str(&String::from_utf16_lossy(string_chars));
+        true
+    }
+    rooted!(in(*cx) let mut data = jsval.get());
+
+    unsafe {
```

Comment:
// 1. Let result be ? [Call](https://tc39.github.io/ecma262/#sec-call)([%JSON.stringify%](https://tc39.github.io/ecma262/#sec-json.stringify), undefined, « value »).

You can add a note that the `if !stringify_result` corresponds to the `?` in the step, because `ToJSON` will throw an exception before returning `false`, which we propagate by returning `Error::JSFailed`.

Reply:
Yes, I was slightly worried about this, in case a caller attempted to return an error other than the `Error::JSFailed` returned here, which could panic because there will be a pending exception set.